jsonapi/unmarshal.go

```diff
@@ -74,428 +72,153 @@ type EditToManyRelations interface {
 	DeleteToManyIDs(name string, IDs []string) error
 }
 
-// Unmarshal reads a JSONAPI map to a model struct
-// target must at least implement the `UnmarshalIdentifier` interface.
-func Unmarshal(input map[string]interface{}, target interface{}) error {
-	var (
-		structType reflect.Type
-		sliceVal   reflect.Value
-		isStruct   bool
-	)
-
-	typeError := errors.New("You must pass a pointer to a UnmarshalIdentifier or slice of it to Unmarshal()")
-
-	// Check that target is a *[]Model
-	ptrVal := reflect.ValueOf(target)
-	if ptrVal.Kind() != reflect.Ptr || ptrVal.IsNil() {
-		return typeError
-	}
-	targetType := reflect.TypeOf(target).Elem()
-
-	if targetType.Kind() != reflect.Slice {
-		// check for a struct or pointer to struct which are also allowed to unmarshal into
-		if targetType.Kind() == reflect.Struct {
-			structType = targetType
-		} else if targetType.Kind() == reflect.Ptr {
-			structType = targetType.Elem()
-		} else {
-			return typeError
-		}
-		sliceVal = reflect.New(reflect.SliceOf(targetType)).Elem()
-		isStruct = true
-	} else {
-		sliceVal = ptrVal.Elem()
-		structType = targetType.Elem()
-		if structType.Kind() == reflect.Ptr {
-			structType = structType.Elem()
-		}
-	}
-
-	if structType.Kind() != reflect.Struct {
-		return typeError
-	}
+var (
+	errInterface         = errors.New("target must implement UnmarshalIdentifier interface")
```

Comment:
I often declare errors that are passed down to the consumer publicly so I can compare against the value. Works better than comparing the string `err.Error() = "..."` because it lets you change/clarify an error in the future without breaking code.